nomos-services/cryptarchia-consensus/src/lib.rs

```diff
@@ -0,0 +1,419 @@
+mod leadership;
+pub mod network;
+use core::fmt::Debug;
+use cryptarchia_ledger::LedgerState;
+use futures::StreamExt;
+use network::NetworkAdapter;
+use nomos_core::block::Block;
+use nomos_core::da::certificate::{BlobCertificateSelect, Certificate};
+use nomos_core::header::{Header, HeaderId};
+use nomos_core::tx::{Transaction, TxSelect};
+use nomos_mempool::{
+    backend::MemPool, network::NetworkAdapter as MempoolAdapter, Certificate as CertDiscriminant,
+    MempoolMsg, MempoolService, Transaction as TxDiscriminant,
+};
+use nomos_network::NetworkService;
+use nomos_storage::{backends::StorageBackend, StorageMsg, StorageService};
+use overwatch_rs::services::life_cycle::LifecycleMessage;
+use overwatch_rs::services::relay::{OutboundRelay, Relay, RelayMessage};
+use overwatch_rs::services::{
+    handle::ServiceStateHandle,
+    state::{NoOperator, NoState},
+    ServiceCore, ServiceData, ServiceId,
+};
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use serde_with::serde_as;
+use std::hash::Hash;
+use thiserror::Error;
+use tokio::sync::oneshot::Sender;
+use tracing::{error, instrument};
+
+#[derive(Debug, Clone, Error)]
+pub enum Error {
+    #[error("Ledger error: {0}")]
+    Ledger(#[from] cryptarchia_ledger::LedgerError<HeaderId>),
+    #[error("Consensus error: {0}")]
+    Consensus(#[from] cryptarchia_engine::Error<HeaderId>),
+}
+
+struct Cryptarchia {
+    ledger: cryptarchia_ledger::Ledger<HeaderId>,
+    consensus: cryptarchia_engine::Cryptarchia<HeaderId>,
+}
+
+impl Cryptarchia {
+    fn tip(&self) -> HeaderId {
+        self.consensus.tip()
+    }
+
+    fn try_apply_header(&self, header: &Header) -> Result<Self, Error> {
+        let header = header.cryptarchia();
+        let id = header.id();
+        let parent = header.parent();
+        let slot = header.slot();
+        let ledger = self.ledger.try_update(
+            id,
+            parent,
+            header.slot(),
```

Comment:
`header.slot()` is stored in the `slot` variable above.

Is header moved in this call to `try_update()`? If not, you could inline all the variables (id()|parent()|slot())